kernel/src/hil/public_key_crypto/keys.rs

```diff
@@ -283,3 +283,98 @@ pub trait RsaPrivKeyMut: PubPrivKeyMut + RsaKeyMut {
+    /// cryptographic operation.
+    ///
+    /// ### `error`:
+    ///
+    /// - `Ok(())`: The key was selected successfully.
+    /// - `Err(())`: The key was selected set successfully.
+    ///   - `ErrorCode::INVAL`: The index was not valid.
+    ///   - `ErrorCode::FAIL`: The key could not be set.
+    fn select_key_done(&self, index: usize, error: Result<(), ErrorCode>);
+}
+
+/// Interface for selecting an active key among the number of available keys.
+///
+/// This interface allows for the implementer to maintain an opaque internal
+/// representation of keys. They may be stored in memory, flash, or in secure
+/// element (where the actual key may not be accessible). Users of this
+/// interface can select which key is active for cryptographic operations. There
+/// is no assumption for implementers of this interface that keys can be added
+/// or changed or that keys can be specified by their actual bytes in a slice.
+///
+/// Selecting a key is asynchronous as it may require communication over a bus
+/// or waiting for an interrupt.
+///
+/// Keys are specified by an index starting at zero and going to
+/// `get_key_count()-1`.
+///
+/// The stored keys can be public or private keys.
+pub trait KeySelect<'a> {
+    /// Return the number of keys that the device can switch among.
+    ///
+    /// Each key must be identifiable by a consistent index.
+    fn get_key_count(&self) -> usize;
+
+    /// Set the key identified by its index as the active key.
+    ///
+    /// Indices start at 0 and go to `get_key_count() - 1`.
+    ///
+    /// This operation is asynchronous and its completion is signaled by
+    /// `select_key_done()`.
+    ///
+    /// ## Return
+    ///
+    /// `Ok()` if the key select operation was accepted. Otherwise:
+    /// - `Err(ErrorCode::INVAL)` if the index is not valid.
+    /// - `Err(ErrorCode::ALREADY)` if the key is already selected.
+    fn select_key(&self, index: usize) -> Result<(), ErrorCode>;
+
+    fn set_client(&self, client: &'a dyn KeySelectClient);
+}
+
+/// Client for setting keys.
+pub trait KeySetClient<const KL: usize> {
+    /// Called when the key has been set.
+    ///
+    /// Returns the previous key if one was set.
+    ///
+    /// ### `error`:
+    ///
+    /// - `Ok(())`: The key was set successfully.
+    /// - `Err(())`: The key was not set successfully.
+    ///   - `ErrorCode::FAIL`: The key could not be set.
+    fn set_key_done(
+        &self,
+        previous_key: Option<&'static mut [u8; KL]>,
+        error: Result<(), ErrorCode>,
+    );
+}
+
+/// Interface for setting keys by a slice.
+///
+/// `KL` is the length of the keys.
+///
+/// Implementers must be able to store keys from a slice. This is most commonly
+/// used for implementations that hold keys in memory. However, this interface
+/// is asynchronous as keys may be stored in external storage or an external
+/// chip and require an asynchronous operations.
+pub trait KeySet<'a, const KL: usize> {
```

Comment:
`KeySet` might be a little confusing, as it could be mistaken with a set of keys. What about `SetKey`? At least for me, a user that has not used the verification infrastructure before, this would have been less confusing.

Reply:
I can get behind that.

Reply:
Renamed.